### PR TITLE
Solution

### PR DESCRIPTION
**Pull request by kadikraman**

### CODE REVIEW

File: components/Animation.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    const progressBar = document.getElementById('progress-bar');
+    progressBar.className = 'off on';
+    setTimeout(() => progressBar.className = 'off', 3000);
+  }
+
+  render() {
+    return (
+      <div>
+        <img src={this.state.url} height="100px"/>
+        <div><button onClick={() => this.getNewCat()}>New random .gif!</button></div>
```
Comment (thomastuts) [inline components/Animation.js]: The `onClick` prop doesn't need to be an arrow function, we can just pass the reference to `this.getNewCat` instead. Make sure it's bound in the constructor, though!

File: components/Animation.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
+const React = require('react');
+
+class Animation extends React.Component {
+
+  // TODO: create a componentWillUpdate() here
```
Comment (thomastuts) [inline components/Animation.js]: We can remove this comment

File: components/Circle.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+const React = require('react');
+
+class Circle extends React.Component {
+
+  // TODO: create a shouldComponentUpdate() here
```
Comment (thomastuts) [inline components/Circle.js]: We can remove this comment

File: components/Circle.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  shouldComponentUpdate(nextProps) {
+    return Boolean(nextProps.color !== this.props.color);
+  }
+
+  render() {
+    const { color } = this.props;
+    console.log(`A ${color} circle was rendered!`)
+    return (
+      <div className={`circle ${color}`}>
```
Comment (thomastuts) [inline components/Circle.js]: We can add the closing `</div>` on the same line as the opening one

File: components/Circle.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
+const React = require('react');
+
+class Circle extends React.Component {
+
+  // TODO: create a shouldComponentUpdate() here
+  shouldComponentUpdate(nextProps) {
+    return Boolean(nextProps.color !== this.props.color);
```
Comment (thomastuts) [inline components/Circle.js]: Not sure why we're casting something to a `Boolean` here, the `!==` will automatically create a boolean value

File: components/Circle.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  }
+
+  render() {
+    const { color } = this.props;
+    console.log(`A ${color} circle was rendered!`)
```
Comment (thomastuts) [inline components/Circle.js]: Missing semicolon

File: components/Circles.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    const newCircles = this.state.circles.map(circle => {
+      if (circle.color === 'red') return { color: 'green' };
+      if (circle.color === 'green') return { color: 'red' };
+      return circle;
+    });
+    console.log("\n💥 Red and green were swapped!");
+    this.setState({ circles: newCircles });
+  }
+
+  constructor(props) {
```
Comment (thomastuts) [inline components/Circles.js]: `constructor` should be the first method in a class definition

File: components/Circles.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    super(props);
+    this.state = {
+      circles: [
+        { color: 'red' },
+        { color: 'yellow' },
+        { color: 'green' },
+      ]
+    }
```
Comment (thomastuts) [inline components/Circles.js]: Missing semicolon

File: components/Circles.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    this.handleClick = this.handleClick.bind(this);
+  }
+
+  render() {
+    return (
+      <div>
+        <button onClick={this.handleClick}>Swap red and green</button>
+        { this.state.circles.map((circle, index) => <Circle color={circle.color} key={index} /> ) }
+      </div>
+    )
```
Comment (thomastuts) [inline components/Circles.js]: Missing semicolon

File: components/Pusheen.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
+const React = require('react');
+
+class Pusheen extends React.Component {
+
+  // TODO: create a componentDidUpdate() here
```
Comment (thomastuts) [inline components/Pusheen.js]: Remove comment

File: components/Pusheen.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    this.resizePusheen();
+  }
+
+  resizePusheen() {
+    let pusheen = document.getElementById("pusheen");
+    pusheen.height = this.state.size;
+    pusheen.width = this.state.size;
+  }
+
+  constructor(props) {
```
Comment (thomastuts) [inline components/Pusheen.js]: `constructor` should be the first method in a class definition

File: components/Pusheen.js [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+class Pusheen extends React.Component {
+
+  // TODO: create a componentDidUpdate() here
+  componentDidUpdate() {
+    this.resizePusheen();
+  }
+
+  resizePusheen() {
+    let pusheen = document.getElementById("pusheen");
```
Comment (thomastuts) [inline components/Pusheen.js]: This can be a `const`

File: components/Pusheen.js [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    this.setState({
+      size: this.state.size - 10
+    });
+  }
+
+  render() {
+    return (
+      <div>
+        <button onClick={() => this.makeBigger()}>Bigger!</button>
+        <button onClick={() => this.makeSmaller()}>Smaller!</button>
```
Comment (thomastuts) [inline components/Pusheen.js]: These two arrow functions can be replaced w/ a simple reference to `this.makeBigger` and `this.makeSmaller` as long as they are bound in the constructor

File: components/Translation.js [hunk 14]
```diff
@@ -0,0 +1,36 @@
+const React = require('react');
+
+class Translation extends React.Component {
+
+  // TODO: create a componentWillReceiveProps here
```
Comment (thomastuts) [inline components/Translation.js]: Remove comment

File: components/Translation.js [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+  componentWillReceiveProps(nextProps) {
+    if (this.props.greetingCode !== nextProps.greetingCode) {
+      this.translate(nextProps.greetingCode);
+    }
+  }
+
+  constructor(props) {
```
Comment (thomastuts) [inline components/Translation.js]: `constructor` should be the first method in a class definition

File: components/Animation.js (hunk 1 above)
Comment (kadikraman) [inline components/Animation.js]: These are equivalent - using an arrow function will work as a bind so you don't have to do it in the constructor.

```
<button onClick={() => this.getNewCat()}
```
is equivalent to
```
<button onClick={this.getNewCat().bind(this)}
```
which is equivalent to
```
getNewCat = () => {
  {---}
}
<button onClick={this.getNewCat()}
```
which is equivalent to
```
constructor{
  this.getNewCat = this.getNewCat.bind(this);
}

<button onClick={this.getNewCat()}
```

I prefer to use arrow functions when possible, but I'll change it to what you suggested if it's what's preferred in the project! :)

File: components/Rating.js [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+  constructor(props) {
+    super(props);
+    this.state = {
+      increasing: false,
+      decreasing: false
+    };
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (this.props.rating < nextProps.rating) {
```
Comment (thomastuts) [inline components/Rating.js]: This is a little verbose and has too much control flow, we can simplify this to:

```js
this.setState({
  increasing: this.props.rating < nextProps.rating,
  decreasing: this.props.rating > nextProps.rating
});
```

File: components/Rating.js [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+    }
+  }
+
+  render() {
+    let trend = 'stable';
+    if (this.state.increasing) {
+      trend = 'increasing';
+    } else if (this.state.decreasing) {
+      trend = 'decreasing';
+    };
```
Comment (thomastuts) [inline components/Rating.js]: No need for a semicolon here

File: components/Survey.js [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+
+  decreaseRating() {
+    this.setState({ rating: this.state.rating - 1 });
+  }
+
+  maintainRating() {
+    this.setState({ rating: this.state.rating });
+  }
+
+  newGreeting() {
```
Comment (thomastuts) [inline components/Survey.js]: This is an old method and can be removed, I think?

File: components/Rating.js (hunk 16 above)
Comment (kadikraman) [inline components/Rating.js]: Haha, that is very true. Not sure what I was thinking 😄

File: components/Animation.js (hunk 1 above)
Comment (thomastuts) [inline components/Animation.js]: Yep those are all valid, but when you bind a method in your constructor, you're not creating a new function in your `render()` method every time it's called, while an inline arrow function will.

Not that performance is super important with tiny apps like these, but it's important to teach the students best practices regarding this stuff so they don't run into any issues themselves later on.

**Pull request merged**